nuxeo-core/nuxeo-core-io/src/main/java/org/nuxeo/ecm/core/io/avro/DocumentModelMapper.java

```diff
@@ -76,25 +68,70 @@ public DocumentModel fromAvro(Schema schema, GenericRecord input) {
 
     @Override
     public GenericRecord toAvro(Schema schema, DocumentModel input) {
-        String logicalType = getLogicalType(schema);
         GenericRecord record = new GenericData.Record(schema);
-        if (DOCUMENT_MODEL.equals(logicalType)) {
-            record.put(PATH, input.getPathAsString());
-            record.put(UUID, input.getId());
-        }
-        for (Field field : schema.getFields()) {
-            if (field.name().equals(UUID) || field.name().equals(PATH)) {
-                continue;
-            }
-            if (field.schema().getType() == Type.RECORD
-                    && (DOCUMENT_MODEL.equals(logicalType) || DOCUMENT_TYPE.equals(logicalType))) {
+        if (Constants.DOCUMENT_MODEL.equals(getLogicalType(schema))) {
+            documentModelToAvro(schema, input, record);
+        } else if (Constants.DOCUMENT_TYPE.equals(getLogicalType(schema))) {
+            for (Field field : schema.getFields()) {
                 record.put(field.name(), service.toAvro(field.schema(), input));
-            } else {
+            }
+        } else {
+            for (Field field : schema.getFields()) {
                 Property p = input.getProperty(service.decodeName(field.name()));
                 record.put(field.name(), service.toAvro(field.schema(), p));
             }
         }
         return record;
     }
 
+    @SuppressWarnings("unchecked")
+    protected DocumentModel documentModelFromAvro(GenericRecord input) {
+        String path = (String) input.get(Constants.PATH);
+        String type = (String) input.get(Constants.PRIMARY_TYPE);
+        String uuid = (String) input.get(Constants.UUID);
+        String parentId = (String) input.get(Constants.PARENT_ID);
+        String repositoryName = (String) input.get(Constants.REPOSITORY);
+        Boolean isProxy = (Boolean) input.get(Constants.IS_PROXY);
+        DocumentRef parentRef = parentId == null ? null : new IdRef(parentId);
+        Set<String> facets = (Set<String>) input.get(Constants.MIXIN_TYPES);
+        DocumentModelImpl doc = new DocumentModelImpl(null, type, uuid, new Path(path), null, parentRef,
+                null, facets, null, repositoryName, isProxy);
+        doc.setIsVersion((Boolean) input.get(Constants.IS_VERSION));
+        doc.prefetchCurrentLifecycleState((String) input.get(Constants.CURRENT_LIFE_CYCLE_STATE));
+        return doc;
+    }
+
+    protected void documentModelToAvro(Schema schema, DocumentModel doc, GenericRecord record) {
+        record.put(Constants.UUID, doc.getId());
+        record.put(Constants.NAME, doc.getName());
+        record.put(Constants.TITLE, doc.getTitle());
+        record.put(Constants.PATH, doc.getPathAsString());
+        record.put(Constants.REPOSITORY, doc.getRepositoryName());
```

Comment:
Why not call the constant (and value) `REPOSITORY_NAME`?

Reply:
because the property set is "repository" but If you prefer REPOSITORY_NAME, so be it